code/controlconfig/controlsconfigcommon.cpp

```diff
@@ -73,6 +74,13 @@ SCP_vector<CCI> Control_config;
 //! Vector of presets. Each preset is a collection of bindings that can be copied into Control_config's bindings. [0] is the default preset.
 SCP_vector<CC_preset> Control_config_presets;
 
+struct LuaHook {
+	bool enabled = false;
+	luacpp::LuaFunction hook;
+	luacpp::LuaFunction override;
+};
+SCP_map<IoActionId, LuaHook> Lua_hooks;
```

Comment:
I could make it an array, but assuming that most of the time very few controls ever have hooks, I opted for the map

Reply:
Maps have O(log(n)) time, making them more performant for a larger set of keys vs. a sequential search in an ordered vector.  This is probably splitting hairs, performance wise, but there is a minimum number of keys that the map will perform better vs. the vector, below that minimum the vector does better.  I have to dig out the graphing calculator to show where that crossover point is.

Reply:
I mean, in an area where n is so small that the O(log n) of the Map is slower than the O(n) of the vector, performance is a non-issue anyways. And so long as that is the case, the map implementation is a lot easier to read IMO.

Reply:
Ah, I might have confused O(log n) it with O(n log n), sorry.

Anyway, we should strive to seek performant solutions for the controls code, and balance it with memory consumption.  Although it may not seem like a big deal with main memory getting progressively larger, the CPU cache isn't growing as quickly, and since controls are a module that should be kept in the cache every frame, it needs some care with persistent data sizes. (puts away soap box).

Reply:
> Ah, I might have confused O(log n) it with O(n log n), sorry.

map::find is O(log n) (since it essentially binary searches instead of going through every element) and is thus faster than iterating through a vector, unless n is very very small. However, maps can be less efficient at using cache, I admit.
As for memory consumption, map needs slightly more, as there is obviously small overhead for each node in the tree. But that is probably neglectable.